google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/ReadWriteTransaction.java

```diff
@@ -105,9 +131,20 @@ class ReadWriteTransaction extends AbstractMultiUseTransaction {
   private final List<Mutation> mutations = new ArrayList<>();
   private Timestamp transactionStarted;
 
-  private static final class RollbackToSavepointException extends Exception {}
+  private static final class RollbackToSavepointException extends Exception {
+    private final Savepoint savepoint;
+
+    RollbackToSavepointException(Savepoint savepoint) {
+      this.savepoint = savepoint;
+    }
+
+    Savepoint getSavepoint() {
+      return this.savepoint;
+    }
+  }
 
   static class Builder extends AbstractMultiUseTransaction.Builder<Builder, ReadWriteTransaction> {
+    private boolean useAutoSavepointsForEmulator;
```

Comment:
should we really need to include `ForEmulator` in the naming? We are not restricting some other client (apart form emulator) from setting this field. Or do we want to annotate this `setUseAutoSavepointsForEmulator` with InternalApi/VisibleForTesting annotations?

Reply:
It is very much intentional, and currently the only way to enable this through any public API is by setting the `autoConfigEmulator=true` connection URL. The reason for this is that the feature is really only intended for the emulator. For real Spanner, this feature should not be used, and instead the client should just rely on the Spanner deciding which transaction to abort or not.
The only reason that we need to override this specifically for the emulator (and specifically in the Connection API), is that some frameworks that depend on JDBC (and similar drivers) require support for executing multiple transactions in parallel using one client thread.
(The `ReadWriteTransaction` class and related builder class are all package-private, and so already internal APIs)